### ISSUE TITLE
need startswith function

### ISSUE THREAD

**Issue by zhangguanzhang** [NikolaLohinski/gonja#10]
```
{{ "test111".startswith('test') }}
```
output
```
True
```

**Comment by zhangguanzhang** [comment]
@NikolaLohinski PTAL

**Comment by NikolaLohinski** [comment]
Hi @zhangguanzhang and thank you for reaching out.

What you are asking for is a pythonism, that would require rewriting the current parser quite a bit, and would open the road to requiring all of python's native method on base types to be implemented as well. Not saying it's not possible, but it's way more work than it looks like.

If you feel like trying, I am happy to take a PR.

Otherwise, you can simply rewrite
```
{{ "test111".startswith('test') }}
``` 
into
````
{{ "test" in ("test123")[0:4] }}
````

And you get the same behavior with pure jinja which currently works.

Cheers

**Comment by zhangguanzhang** [comment]
> Hi @zhangguanzhang and thank you for reaching out.
> 
> What you are asking for is a pythonism, that would require rewriting the current parser quite a bit, and would open the road to requiring all of python's native method on base types to be implemented as well. Things like
> 
> ```
> {{ 2.0.hex() }}
> ```
> 
> Output
> 
> ```
> 0x1.0000000000000p+1
> ```
> 
> Not saying it's not possible, but it's way more work than it looks like. If you feel like trying, I am happy to take a PR.
> 
> Otherwise, you can simply rewrite
> 
> ```
> {{ "test111".startswith('test') }}
> ```
> 
> into
> 
> ```
> {{ "test" in "test123"[0:4] }}
> ```
> 
> (you'd need v2.0.2 for this, cause I just fixed a bug regarding accessing raw string literals)
> 
> And you get the same behavior with pure jinja which currently works.
> 
> Cheers

thanks , I really need these template string methods:
```
{% set my_string = "abc" %}

{{ my_string.startswith('a') }}
{{ my_string.endswith('c') }}
{{ my_string.upper() }}
{{ my_string.lower() }}
{{ my_string.replace('a', 'x') }}
{{ my_string.strip() }}
{{ my_string.split('b') }}
```

and I think there is a bug with `in`:

```
{{ "test" in "test111"[0:4] }}
{{ "1" in test111"[-1] }}
```
output:
```
{{ "1" in test111"[-1] }}': '}}' expected here (Line: 0 Col: 0, near "[-1] }}")
```

**Comment by NikolaLohinski** [comment]
Hey @zhangguanzhang 

Can you upgrade to v2.0.2 ? Negative indices are supported since 2.0.0 and the bug regarding raw string literals slicing should be fixed in that bug fix version. And if not, I'll look into it. 

Regarding the initial topic and the methods on native Python objects, I took a quick glance and it might be possible to implement, but as I said it's quite a bit of work, and requires reimplementing quite a few Python functions in Go 🤐
I won't have time in the next weeks and maybe months, but I'll keep that in mind going forward, rest assured :)

**Comment by zhangguanzhang** [comment]
@NikolaLohinski 
Negative indices works fine after upgrade to v2.0.2 .👏
I have looked for many other golang jinjia libraries, but none of them seem to be active. I will pay attention to the subsequent updates of this repo.🎉🎉🎉

**Comment by NikolaLohinski** [comment]
@Zatte @zhangguanzhang I just merged https://github.com/NikolaLohinski/gonja/pull/11

and released `v2.1.0` with a few methods implemented for all basic types amongst which you will find `startswith` for strings. 

Feel free to make PRs to add whatever other methods you need, I'll be happy to review them and help you sort out stuff. 

If you do so, please make sure to document stuff in `docs/methods.md` and write tests in `tests/integration/...`.  to run with [ginkgo](https://onsi.github.io/ginkgo/) (I have yet to write a contribution guide so bear with me here :innocent: ).

I am going to close this issue cause the initial reques is now fulfilled, but feel free to reopen it if you find it necessary.

Cheers

**Comment by zhangguanzhang** [comment]
nice work, It works fine for me. Cheers

**Comment by Zatte** [comment]
@NikolaLohinski, based on this i played around with the 1:1 mapping of python std lib methods.  https://github.com/Zatte/gopylibs/tree/main/pystring

Most of the the methods are implemented, including "format()", (though most methods are only QA:ed with chatgpt generated table tests. The format directive is bit more battle tested) I have a lot of cleanup to do but got stuck with one design decision and wanted to hear your thoughts on a few before I continue. 

**Error Handling**
1) Errors - in python it is common to raise errors, not so much in go. Designing functions around returning errors complicates chaining. I see 2 options. 
1a) return errors and let gonja handle the chaining - Most go canonical.
1b) Let `PyString` and other `Py`-types becomes monadic-like and use a final getter is needed to receive the boxed value or error - More overhead, not as intuitive for go-users. 

**Boxed / Aliased types**
Right now all `PyString` methods comes in two versions, one operating on native string types (function) and one on `PyString` (type method) which allows chaining. The later sometime returns `bool` (eg. `.statswith`) and `int` (e.g. ``count()``). I think it could make sense to instead return `PyBool` and `PyInt` to allow for `fmt.Stringer` and method chaining of those types as well. This would complicate the monadic-like solution above due to golangs type system. 

**TLDR; seeking input on two questions.** 
- How to handle errors but to still have a pythonic UX. 
- Returning boxed/aliased py-types or native types.

**Comment by NikolaLohinski** [comment]
Hi @Zatte and thanks for taking some time to look into this.

1. I would stick with canonical go, having in mind that the closest equivalent to python exception raising is a regular error returning in Go (if we don't consider using panics ofc, but that just feels wrong). We're not trying to be pythonic, but provide a go implementation of a subset of python. So I believe it should stay true to Go. 
2. I would return Boxed types, to simplify chaining, and provide casting functions for those types (basically `.ToString()` for `PyString`, `.ToInt()` for `PyInt` etc... ). It just feels simpler in the context of a library, especially for chaining stuff. 

What do you have in mind for implementing the methods that mutate the type such as `reverse`  in lists for example ? I implemented this one with reflection in https://github.com/NikolaLohinski/gonja/blob/master/builtins/methods/list.go#L24 , but I am curious to see what you have in mind.

**Comment by Zatte** [comment]
1) - Thanks; and i agree. 
2) - Will probably go with `PyString() PyString` and `String() string` signatures so that also native types can be recovered.

Given that gonja is already heavily using reflection i don't see this as a problem and for generic container types I'm not even sure it could be solved any other way; The actual swapping algorithm can probably be more efficient but that i think is secondary at this stage.

**Comment by Zatte** [comment]
@NikolaLohinski Sorry for the delay, been swamped with (paid) work 

Dropped a placeholder PR here https://github.com/NikolaLohinski/gonja/pull/15 to either allow someone else to continue carrying the torch or at least start the feedback process.
